samples/remote_attestation/common/dispatcher.cpp

```diff
@@ -182,10 +182,177 @@ int ecall_dispatcher::verify_report_and_set_pubkey(
+                      "of the encrypted key");
+
+        // Step 4 - Sign this SHA hash with my enclave's private key
+        if (m_crypto->sign(digest, 32, signature, &signature_size) != 0)
+        {
+            goto exit;
+        }
+
+        TRACE_ENCLAVE(
+            "enclave: establish_secure_channel: signature_size = %ld",
+            signature_size);
+
+        if ((encrypted_key_size != 256) || (signature_size != 256))
+        {
+            TRACE_ENCLAVE("enclave: establish_secure_channel: failed as "
+                          "encrypted data size or signature size is not 256");
+        }
+        // Step 5 - Send digest plus signature to the other enclave
+        memcpy(host_buf, encrypted_key_buf, total_size);
+
+        TRACE_ENCLAVE(
+            "enclave: establish_secure_channel: total_size = %ld", total_size);
+        *key = host_buf;
+        *key_size = total_size;
+    }
+    else
+    {
+        goto exit;
+    }
+    ret = 0;
+exit:
+    return ret;
+}
+
+/* Encrypted key_buf should contain encrypted key followed by signature, each of
+ * 256 bits
+ */
+int ecall_dispatcher::acknowledge_secure_channel(
+    uint8_t* encrypted_key_buf,
+    size_t encrypted_key_size)
+{
+    int ret = 1;
+    uint8_t* data;
+    size_t data_size = 32;
+    uint8_t digest[32];
+    int rc;
+
+    /* Steps --
+     *   1) Verify Signature; if good proceed to step 2
+     *   2) Decrypt the key using your own private key
+     *   3) Now use this key with sequence number to communicate further
+     */
+    unsigned char* signature = &encrypted_key_buf[256];
+    size_t signature_size = 256;
+
+    data = m_enclave_config->sym_key;
+    m_enclave_config->sequence_number = 0;
+
+    if (m_initialized == false)
+    {
+        TRACE_ENCLAVE("ecall_dispatcher initialization failed.");
+        goto exit;
+    }
+
+    if (m_crypto->sha256(encrypted_key_buf, 256, digest) != 0)
+    {
+        goto exit;
+    }
+
+    rc = m_crypto->verify_sign(
+        m_crypto->get_the_other_enclave_public_key(),
+        digest,
+        32,
+        signature,
+        signature_size);
+    if (rc != 0)
+    {
+        TRACE_ENCLAVE(
+            "enclave: acknowledge_secure_channel: signature "
+            "verification failed with %x\n",
+            rc);
+        goto exit;
+    }
+
+    TRACE_ENCLAVE("enclave: acknowledge_secure_channel: signature verified ok");
+
+    if (m_crypto->decrypt(encrypted_key_buf, 256, data, &data_size))
+    {
+        TRACE_ENCLAVE(
+            "enclave: acknowledge_secure_channel: extracted symmetric key size "
+            "= %ld",
+            data_size);
+    }
+    else
+    {
+        TRACE_ENCLAVE("enclave: acknowledge_secure_channel: symmetric key "
+                      "decrypt failed");
+        goto exit;
+    }
+
+    ret = 0;
+exit:
+    return ret;
+}
+
 int ecall_dispatcher::generate_encrypted_message(uint8_t** data, size_t* size)
 {
```

Comment:
This routine seems to assume that a secure channel was already established, which might not be true in all cases.
m_initialized could be re-purpose for it.

Reply:
Yes, good point. Will incorporate.

Reply:
Established a state machine to make this possible. Definitely cleaner.